index.ts

```diff
@@ -432,28 +446,236 @@ export class CapabilitiesController extends BaseController<any, any> implements
+  ): void {
+
+    // assert caveat is valid
+    if (!this.validateCaveats([caveat])) {
+      throw internalError({
+        message: 'Invalid caveat param. Must be a valid caveat object.',
+        data: caveat,
+      });
+    }
+
+    const perm = this._getPermissionForCaveat(
+      domainName, methodName
+    );
+
+    const newCaveats = (perm.caveats && [ ...perm.caveats ]) || [];
+
+    this._validateAndUpdateCaveats(
+      domainName, methodName, caveat, newCaveats, perm
+    );
+  }
+
+  /**
+   * Overwrites the caveat with the given name for the permission
+   * corresponding to the given domain and method. Throws if the domain
+   * or method are unrecognized, or if a caveat with the given name doesn't
+   * exist.
+   * 
+   * @param {string} domainName - The grantee domain.
+   * @param {string} methodName - The name of the method identifying the permission.
+   * @param {IOcapLdCaveat} caveat - The new caveat for the permission.
+   */
+  updateCaveatFor (
+    domainName: string,
+    methodName: string,
+    caveatName: string,
+    caveatValue: any
+  ): void {
+
+    if (!caveatName || typeof caveatName !== 'string') {
+      throw internalError({
+        message: 'Invalid caveat param. Must specify a string name.',
+        data: caveatName,
+      });
+    }
+
+    const perm = this._getPermissionForCaveat(
+      domainName, methodName
+    );
+
+    // get target caveat
+    const caveat = perm.caveats && perm.caveats.find(
+      c => c.name === caveatName
+    );
+
+    // copy over all caveats except the target
+    const newCaveats: IOcapLdCaveat[] = []
+    perm.caveats && perm.caveats.forEach(c => {
+      if (c.name !== caveatName) {
+        newCaveats.push(c);
+      }
+    });
+
+    // assert that the target caveat exists
+    if (!caveat || !perm.caveats) {
+      throw internalError({
+        message: 'No such caveat exists for the relevant permission.',
+        data: caveatName
+      });
+    }
+
+    if (typeof caveat.value !== typeof caveatValue) {
+      throw internalError({
+        message: 'New caveat value is of different type than original.',
+        data: { caveat, newValue: caveatValue }
+      });
+    }
+
+    caveat.value = caveatValue;
+
+    this._validateAndUpdateCaveats(
+      domainName, methodName, caveat, newCaveats, perm
+    );
+  }
+
+  /**
+   * Internal function used in addCaveatFor and updateCaveatFor.
+   */
+  private _getPermissionForCaveat (
+    domainName: string,
+    methodName: string
+  ): IOcapLdCapability {
+
+    // assert domain already has permission
+    const perm = this.getPermission(domainName, methodName);
+    if (!perm) {
+      throw internalError({
+        message: 'No such permission exists for the given domain.',
+        data: { domain: domainName, method: methodName },
+      });
+    }
+
+    return perm;
+  }
+
+  /**
+   * Internal function used in addCaveatFor and updateCaveatFor.
+   */
+  private _validateAndUpdateCaveats (
```

Comment:
I think this method does two different things that feel distinct:
(1) This method combines concatenates a caveat with an array of caveats and then checks if the resultant array is valid
(2) This method overwrites a permissions caveats with a new array of caveats

I think it would make sense for this method to just receive a `newCaveats` array, but not a `caveat` and simple be responsible for overwriting the permissions caveats with the received `newCaveats`.  Validation could still happen here before overwrite, but it would become the caller's responsibility to construct the `newCaveats` array.

Alternatively, this method could include the following logic:
```
const perm = this._getPermissionForCaveat(
      domainName, methodName
    );

const newCaveats = (perm.caveats && [ ...perm.caveats ]) || [];
```
and therefore not need to receive either `newCaveats` or a `perm`. This would make the use of the method more specific, and less dependent on its callers. The only reason not to do this, I think, would be whether there is a need to overwrite a permission with an entirely new permission that is not a modification of the existing permission (which is different from the existing use case of this method).

Reply:
Made the construction of `newCaveats` the responsibility of the callers, and removed `caveat` as an argument. Would've liked to remove `perm` as an argument, but there's no way to construct `newCaveats` in the way we intend without getting the perm first, so can't justify removing it.